### ISSUE TITLE
[Presets: React + Typescript] Interface with generic method fails transpilation

### ISSUE THREAD

**Issue by lukeggchapman** [babel/babel#7504]
### Input Code

This is only an issue when using both the react and typescript presets.
```js
interface I {
    <T>(p: T): T
}
```

[Repl Link](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=JYOwLgpgTgZghgYwgAgJLIN4Chm-QHgBUA-ACgAcAuZQgSmsKwF8sg&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=true&fileSize=false&lineWrap=true&presets=react%2Ctypescript&prettier=false&targets=&version=7.0.0-beta.38&envVersion=)

### Babel/Babylon Configuration (.babelrc, package.json, cli command)

```js
{
    "presets": [
        "@babel/react",
        "@babel/typescript"
    ],
}
```

### Expected Behavior

It should not throw and error.

### Current Behavior

It throws:
```
{ SyntaxError: ./src/test.tsx: Unexpected token (2:4)

  1 | interface I {
> 2 |     <T>(p: T): T
    |     ^
  3 | }
  4 |
```

### Possible Solution

Similar issue with generic methods: #6665
Similar fix for generic methods: #7225
Duplicate but closed Issue: #7158 

### Your Environment

| software         | version(s)
| ---------------- | -------
| Babel            |  7.0.0-beta.40
| preset-react            |  7.0.0-beta.40
| preset-typescript           |  7.0.0-beta.40
| node             | 8.9.0
| npm              | 5.5.1
| Operating System | OSX

**Comment by jon-lp** [comment]
@lukeggchapman FWIW moving the function interface declaration into a separate file.d.ts file seemed to make everything happy:
– Webpack/Babel successfully compiles 
– tsc type check works with that interface
– VSCode sees/follows it and has no errors/warnings
hope that helps!

**Comment by Tsury** [comment]
Also encountering this.

**Comment by danez** [comment]
Fixed by #9058
